### PR TITLE
Container view controller

### PR DESCRIPTION
**Pull request by marijnschilling**
Set up the container view controller with
- a top viewcontroller
- a navigation bar
- a scrollview containing the bottom view controllers
- a gesture recogniser going from top to bottom floor 

- check out that neato layout code in the `LiftBottomScrollView` to layout as much viewControllers on the scrollview as you wish 💅

Future plans:
- I want to add a drag with a threshold to activate the transition
- The scrolling on the scrollView should snap to the frames

navigation

### CODE REVIEW

File: Sources/LiftBottomScrollView.swift [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import UIKit
+
+class LiftBottomScrollView: UIScrollView {
```
Comment (3lvis) [inline Sources/LiftBottomScrollView.swift]: I think you can remove the `Lift` suffix in internal classes since the context of usage is quite clear.

Comment (3lvis) [reply Sources/LiftBottomScrollView.swift]: I meant prefix

File: Demo/AppDelegate.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        secondViewController.view.backgroundColor = .white
+        secondViewController.view.layer.borderWidth = 5.0
+        secondViewController.view.layer.borderColor = UIColor.kronBlue().cgColor
+
+        let thirdViewController = UIViewController()
+        thirdViewController.view.backgroundColor = .white
+        thirdViewController.view.layer.borderWidth = 5.0
+        thirdViewController.view.layer.borderColor = UIColor.kronPink().cgColor
+
+        let controller = LiftNavigationController(topViewController: topViewController, bottomViewControllers: [firstViewController, secondViewController, thirdViewController])
```
Comment (3lvis) [inline Demo/AppDelegate.swift]: Nice start! Where are you planning to get the labels from?

File: Demo/Color+Helpers.swift [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import UIKit
+
+public extension UIColor {
```
Comment (3lvis) [inline Demo/Color+Helpers.swift:3]: I think is fine if you add SweetUIKit to the Demo, the Library though should be dependency-free (if possible).

Comment (3lvis) [reply Demo/Color+Helpers.swift:3]: Although, probably better to drop hex at all from this project.

File: Demo/Color+Helpers.swift [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    
+    class func kronGreen() -> UIColor {
+        return UIColor(hex: "58D0A8")
+    }
+
+    class func kronBlue() -> UIColor {
+        return UIColor(hex: "66A5FF")
+    }
+
+    class func kronPink() -> UIColor {
```
Comment (3lvis) [inline Demo/Color+Helpers.swift]: Might be good to remove "Kron" from here.

Comment (elland) [reply Demo/Color+Helpers.swift]: Why are colours bundled with the component? 🤔

File: Sources/LiftBottomScrollView.swift [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    func addBottomViewControllersAndConstraints(_ bottomViewControllers: [UIViewController]) {
+        for (index, viewController) in bottomViewControllers.enumerated() {
+            viewController.view.translatesAutoresizingMaskIntoConstraints = false
+            self.addSubview(viewController.view)
+
+            let firstViewController = index == 0
+            let lastViewController = index == bottomViewControllers.count - 1
+            let middleViewController = !firstViewController && !lastViewController
```
Comment (3lvis) [inline Sources/LiftBottomScrollView.swift]: These booleans could read much better if they were `isFirstViewController`, `isLastViewController` and so on.

**Review by 3lvis — CHANGES_REQUESTED** [review]

File: Demo/Color+Helpers.swift (hunk 4 above)
Comment (3lvis) [reply Demo/Color+Helpers.swift]: @Elland this is just for the demo project.

Comment (elland) [reply Demo/Color+Helpers.swift]: 👌

File: Sources/TopController.swift [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import UIKit
+
+class TopController: UIViewController {
```
Comment (3lvis) [inline Sources/TopController.swift]: Are we gonna use this class?

File: Sources/LiftNavigationController.swift [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        if recognizer.direction == .up {
+            self.heightAnchor?.constant = -(self.view.bounds.height - LiftNavigationController.navigationBarHeight)
+            self.gestureRecognizer.direction = .down
+        } else {
+            if recognizer.direction == .down {
+                self.heightAnchor?.constant = 0.0
+                self.gestureRecognizer.direction = .up
+            }
+        }
+        UIView.animate(withDuration: 0.2, delay: 0.0, options: [.curveEaseOut], animations: {
```
Comment (3lvis) [inline Sources/LiftNavigationController.swift]: I usually add `.allowUserInteraction, .beginFromCurrentState` to animations, makes them smoother.

**Comment by 3lvis** [comment]
Looks like a solid first iteration, @marijnschilling 💐

File: Demo/AppDelegate.swift (hunk 2 above)
Comment (marijnschilling) [reply Demo/AppDelegate.swift]: I plan on using the `viewController.title` for this

File: Demo/Color+Helpers.swift (hunk 3 above)
Comment (marijnschilling) [reply Demo/Color+Helpers.swift:3]: I like the hex though! I can get it straight from the designs of the project. I's nice to have the demo applied to the theme.

But adding SweetUIKit via carthage i get:
```
Library not loaded: @rpath/SweetUIKit.framework/SweetUIKit
Referenced from: /var/containers/Bundle/Application/B7568C16-DAF3-4C29-9E48-CC36ADA33253/Demo.app/Demo
Reason: image not found
```

Comment (elland) [reply Demo/Color+Helpers.swift:3]: You have to also copy the frameworks on build using a `copy files` step. 👍 

Also remember to put the dependency inside `Cartifle.private` 😄

Comment (marijnschilling) [reply Demo/Color+Helpers.swift:3]: I don't know any of that :) What is the `copy files` step? And what is `Cartfile.private`

Comment (3lvis) [reply Demo/Color+Helpers.swift:3]: Is probably better to keep it as lean as possible, using plain RGB colors for the demo.

**Comment by marijnschilling** [comment]
@Elvis Did all the things!

**Comment by 3lvis** [comment]
@marijnschilling can you delete the Cartfile and the Carthage folder?

**Comment by marijnschilling** [comment]
How does that work than? Do I need to add a new cartfile every time?

**Comment by 3lvis** [comment]
I asked if you could delete it because I thought you weren't using it, but if you are I think it makes sense to keep it.

**Comment by 3lvis** [comment]
Personally, I would prefer to not use any dependencies at all in a library as small as this one, though.

**Comment by marijnschilling** [comment]
I dont know, i can also add the helper method to create a color form a hex as I had before :)

**Comment by elland** [comment]
It seems very odd to use hex strings for colour for a demo project like this. I'd just use literals or `UIColor()`, the simplest most native possible and move on. 👍

**Comment by marijnschilling** [comment]
I want to give it a bit of the look and feel of the design of the product though, so I do need this method.

**Comment by elland** [comment]
But this is a component tho, there is no product. 😄

**Comment by elland** [comment]
You can also just lift those colours using xScope. 👍

**Comment by marijnschilling** [comment]
There is though. I'm making this to show to the client. I'm just adding this method for now, to much work to lift the colours and everything (although the time spend discussing this would cover it)  Might remove it later on! Now i just need fast iteration for now.

**Comment by 3lvis** [comment]
@marijnschilling just keep your little class instead of SweetUIKit then!

**Comment by marijnschilling** [comment]
💥

**Review by 3lvis — APPROVED** [review]

**Comment by 3lvis** [comment]
Ship it!

**Comment by elland** [comment]
🐑

**Pull request merged**